v2/bundle/spiffebundle/bundle.go

```diff
@@ -3,175 +3,411 @@ package spiffebundle
+			return nil, spiffebundleErr.New("missing use for key entry %d", i)
+		}
+	}
+
+	return bundle, nil
 }
 
 // FromX509Bundle creates a bundle from an X.509 bundle.
 func FromX509Bundle(x509Bundle *x509bundle.Bundle) *Bundle {
-	panic("not implemented")
+	if x509Bundle != nil {
+		return &Bundle{
+			trustDomain: x509Bundle.TrustDomain(),
+			x509Roots:   x509Bundle.X509Roots(),
+		}
+	}
+	return &Bundle{}
 }
 
 // FromJWTBundle creates a bundle from a JWT bundle.
 func FromJWTBundle(jwtBundle *jwtbundle.Bundle) *Bundle {
-	panic("not implemented")
+	if jwtBundle != nil {
+		return &Bundle{
+			trustDomain: jwtBundle.TrustDomain(),
+			jwtKeys:     jwtBundle.JWTKeys(),
+		}
+	}
+	return &Bundle{}
+}
+
+// FromX509Roots creates a bundle from X.509 certificates.
+func FromX509Roots(trustDomain spiffeid.TrustDomain, x509Roots []*x509.Certificate) *Bundle {
+	return &Bundle{
+		trustDomain: trustDomain,
+		x509Roots:   x509Roots,
+	}
+}
+
+// FromJWTKeys creates a new bundle from JWT public keys.
+func FromJWTKeys(trustDomain spiffeid.TrustDomain, jwtKeys map[string]crypto.PublicKey) *Bundle {
+	return &Bundle{
+		trustDomain: trustDomain,
+		jwtKeys:     jwtKeys,
+	}
 }
 
 // TrustDomain returns the trust domain that the bundle belongs to.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	return b.trustDomain
 }
 
 // X509Roots returns the X.509 roots in the bundle.
 func (b *Bundle) X509Roots() []*x509.Certificate {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	return b.x509Roots
 }
 
 // AddX509Root adds an X.509 root to the bundle. If the root already
 // exists in the bundle, the contents of the bundle will remain unchanged.
-func (b *Bundle) AddX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) AddX509Root(x509Root *x509.Certificate) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	for _, r := range b.x509Roots {
+		if x509util.CertsEqual(r, x509Root) {
+			return
+		}
+	}
+
+	b.x509Roots = append(b.x509Roots, x509Root)
 }
 
 // RemoveX509Root removes an X.509 root from the bundle.
-func (b *Bundle) RemoveX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) RemoveX509Root(x509Root *x509.Certificate) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	for i, r := range b.x509Roots {
+		if x509util.CertsEqual(r, x509Root) {
+			b.x509Roots = append(b.x509Roots[:i], b.x509Roots[i+1:]...)
+			return
+		}
+	}
 }
 
 // HasX509Root checks if the given X.509 root exists in the bundle.
 func (b *Bundle) HasX509Root(root *x509.Certificate) bool {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	for _, r := range b.x509Roots {
+		if x509util.CertsEqual(r, root) {
+			return true
+		}
+	}
+	return false
 }
 
 // JWTKeys returns the JWT keys in the bundle, keyed by key ID.
 func (b *Bundle) JWTKeys() map[string]crypto.PublicKey {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	return b.jwtKeys
```

Comment:
We should probably copy the map so we don't accidentally introduce false sharing.